client/client.go

```diff
@@ -1184,3 +1162,45 @@ func (c *Client) addAlloc(alloc *structs.Allocation) error {
 	c.allocLock.Unlock()
 	return nil
 }
+
+// setupConsulClient creates a ConsulService
+func (c *Client) setupConsulClient() error {
+	cfg := consul.ConsulConfig{
+		Addr:      c.config.ReadDefault("consul.address", "127.0.0.1:8500"),
+		Token:     c.config.Read("consul.token"),
+		Auth:      c.config.Read("consul.auth"),
+		EnableSSL: c.config.ReadBoolDefault("consul.ssl", false),
+		VerifySSL: c.config.ReadBoolDefault("consul.verifyssl", true),
+	}
+
+	cs, err := consul.NewConsulService(&cfg, c.logger)
+	c.consulService = cs
+	return err
+}
+
+// syncConsul removes services of tasks which are no longer in running state
+func (c *Client) syncConsul() {
+	sync := time.After(consulSyncInterval)
+	for {
+		select {
+		case <-sync:
+			var runningTasks []*structs.Task
+			for _, ar := range c.allocs {
```

Comment:
You need to lock `c.allocLock` when accessing `c.allocs`. I would grab the alloc runners in one shot with the lock, then iterate over them. This way you don't have to hold the lock the full length of the sync. It blocks availability of starting/stopping/updating allocs